### DISCUSSION TITLE
Quick question on redefining blocks/nodesets/sidesets with MOOSE

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#25190] [Q&A Meshing] [answered]
I use gmsh primarily for meshing but discovered that with local refinement some "physical groups" aka block names disappear because they are not tied to the locations in space. So I'm looking for a fix for that.

I saw someone defining a boundary with MOOSE on another question and was wondering if that would be useful for my case. For example, if I have an existing mesh with no blocks defined (or "wrong" blocks defined), can I use some inbuilt MOOSE functions to redefine them?

For example, I found "SubdomainBoundingBoxGenerator" which, if I understand correctly, takes the lower left and upper right coordinates and assigns all elements within the cubic/3d rectaungular "box" defined by those points to a new subdomain ID you can define?

Can you also do the same for more complex geometries? Say, 3D triangles? 

What about for surfaces? In one of the input file examples, I saw "ParsedGenerateSideset" where it seems like the area was defined with a parsed mathematical function. So, for example, would a plane rectangular surface on the top be "(z=0) & (x>0) & (x<x_len) & (y>0) & (y<ylen)"?

Also, I found "ParsedGenerateSideSet" accidentally while looking through an example input file. I can't seem to find a list of all available options here, I tried using the url and went to "https://mooseframework.inl.gov/source/meshgenerators/" to find a list of them but got a 403 Forbidden error.

Another related question is on mesh refinement. Can you use the parsed generate sideset (or something else) to define a "block", and then use the mesh refinement option to refine that block alone?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Blocks / physical groups should not disappear in the first place. Especially not with mesh refinement.

You can use MOOSE redefine blocks in a number of ways. Please have a look at the subdomain generating mesh generators:
https://mooseframework.inl.gov/syntax/Mesh/index.html

You can use parsed expressions for more complex geometries. You can also just create your own mesh generator to do it.

A rectangular plane on top could be:
```
"(z=${top}) & (x>0) & (x<x_len) & (y>0) & (y<ylen)"?
```
but be careful with the equality of floating point numbers. Use this instead:
```
"(z>${top}-1e-8) & (z<${top}+1e-8) & (x>0) & (x<x_len) & (y>0) & (y<ylen)"?
```

You can use the RefineBlockGenerator to refine a single block

Guillaume

**Reply by richmondodufisan** [reply]
Awesome, thanks! And yeah the problem with blocks disappearing is an issue with gmsh, not MOOSE. (https://gitlab.onelab.info/gmsh/gmsh/-/issues/1786)

Also, another question: If existing blocks are already defined, and there is partial overlap with newly created blocks in MOOSE, will the MOOSE blocks override the former? Say, an existing block from an external meshing software is a sphere partially cutting into a rectangular box, and the sphere is defined as a block. If you assign an ID in MOOSE to all elements within the rectangular box (thus, including parts of the sphere), how does it handle that? Can you tell it to ignore original block IDs? 

Also, will you also be required to assign an ID to the part of the sphere that isn't within the rectangular block or will it just be assumed that any block not explicitly defined belongs to the whole mesh? (i.e, any kernels and materials that don't have the block specified will act on that part)?

**Reply by GiudGiud** [reply]
Yes the later generators take precedence for renaming blocks unless there is a specific parameter to override that (like exclude_subdomain_id)
